RedHatInsights/rhsm-subscriptions
# Pull OpenShift Metrics from Prometheus

This PR adds the ability to query a Prometheus instance for openshift metrics and persist the metrics as Events to the rhsm-subscriptions database.

A metering package was created to contain all things metering. It is structured in a way that should allow adding other prometheus based metering fairly easily in the future.
      
The general flow for metrics updates:
- process initiated via JMX or job run.
- a per account task message is sent to a kafka topic with the details of the metrics update that is to be performed.
- when a message is received by the worker, the metrics are pulled from prometheus via the
  Prometheus client, translated to an Event and stored in the database.

Profiles
--------
I tried to make profiles and configuration for the metering components a little more clear and predictible.
I deviated a little from the existing pattern because of some of the issues that I ran into. I'm
OK with re-aligning if it is an issue.

Three new profiles were added:
- _openshift-metering-jmx_: Loads the JMX endpoints for pulling metrics
- _openshift-metering-job_: Launches the metrics gathering process and exits.
- _openshift-metering-worker_: Monitors the openshift metrics queue for tasks and performs the actual work on a per account basis.

**NOTE:** The JMX endpoints are activated along with the openshift-metering-worker profile but can be deployed on their own. This is useful if you'd like to queue up a bunch of messages to be processed once the worker comes online.
                             
                             
Prometheus HTTP Client
------------------
A prometheus client is being generated by an openapi schema that defines both the QueryAPI
as well as the RangeQueryAPI. At this point in time, the prometheus project does not have
an openapi schema for there API so we had to create our own.


Querying Prometheus Metrics
---------------------------
Metrics are being pulled from prometheus following a similar pattern as the Thanos Querier web UI, in that
given a time range, end date and a resolution/step, we get metrics back for a given query. The metering component,
allows tweaking the time range and resolution via configuration.

There are a few key settings and parameters that should be mentioned:
- **End Date:** The last time point to include a matching metric.
- **Time Range** (in minutes): Defines a period of time before the 'end date' that defines the range of the query. For example, a range of 60 minutes would yeild: START_DATE = END_DATE - 60 minutes
- **Resolution/step**: Defines how the metric data should be partitioned/combined. For example, if you query with a range of an hour, you could specify a resolution of 600 seconds (10 mins) and the results would be the max sample taken (if it exists) for each 10 minute block in the 1h time range.

The time range and resolution values are configurable to allow us to customize the results we get. By default, we use a 1 hour time range and a 1 hour resolution.


Testing This PR
---------------
**Reach out if you need the details for the prometheus (telemeter) instance as I didn't want to make the details public. I used one from DataHub**

1. Determine the URL and Auth token of the prometheus server you are connecting to.
2. Launch a local kafka instance.
3. Launch the openshift metering worker (JMX bean is loaded by default with the worker profile). Be sure to substitute
   your values for <YOUR_PROMETHEUS_INSTANCE_URL> and <YOUR_AUTH_TOKEN>.

```bash
$ DEV_MODE=true RHSM_SUBSCRIPTIONS_METERING_PROMETHEUS_METRIC_OPENSHIFT_RANGE_IN_MINUTES=180 PROM_AUTH_TOKEN=<YOUR_AUTH_TOKEN> PROM_URL=<YOUR_PROMETHEUS_INSTANCE_URL> ./gradlew clean bootRun --args="--spring.profiles.active=openshift-metering-worker,kafka-queue"

```
**NOTE**: I'm setting the RANGE_IN_MINUTES configuration value to 3h because the data on datahub's telemeter instance seems to be a few hours behind. This makes sure that any metric's gatered from NOW (any endpoint that does not require an end_date parameter) will yeild results.

4. Visit: http://localhost:8080/actuator/hawtio/jmx/ and navigate to the org.candlepin.subscriptions.metering.jmx where you should  find the OpenShiftJmxBean.
5. Test all of the operations to ensure that data can be retrieved. Event data should be persisted in the DB when run.
   * Turn debugging on to see more info
   * Use account 6274079 as this should yeild data.
   * Use an end date of 2021-01-10T09:12:28Z paired with a range of 3600
6. Test that the job runs (ensure that you have an opted in account)
```bash
DEV_MODE=true RHSM_SUBSCRIPTIONS_METERING_PROMETHEUS_METRIC_OPENSHIFT_RANGE_IN_MINUTES=180 PROM_AUTH_TOKEN=$(echo $OS_AUTH_TOKEN) PROM_URL="https://telemeter-lts.datahub.redhat.com/api/v1" ./gradlew clean bootRun --args="--spring.profiles.active=openshift-metering-job,kafka-queue"
```

**NOTE:** This can also be tested without Kafka, but it is better that we test with Kafka enabled since that is how we will be deploying it.
